roles/common/tasks/networking.yml

```diff
@@ -79,28 +79,53 @@
 #Each host can belong to only one undercloud_cidr, so there will be only one nw_gw set per host.
 - name: set gateway address for multi subnet
   set_fact: nw_gw={{ item.gateway }}
-  when: (update_routes|default('false')|bool) and ((ansible_default_ipv4.network ==  ( item.cidr | ipaddr('network') )))
   with_items: "{{ undercloud_cidr }}"
+  when:
+    - update_routes|default('false')|bool
+    - ansible_default_ipv4.network == (item.cidr | ipaddr('network'))
 
 #Add or replace a static route to all undercloud_cidr subnets via the nw_gw of the current subnet.
 - name: add static route for multi subnet
   command: "/sbin/ip route replace {{ item.cidr }} via {{ nw_gw }}"
-  when: (update_routes|default('false')|bool) and (not (ansible_default_ipv4.network ==  ( item.cidr | ipaddr('network') )))
   with_items: "{{ undercloud_cidr }}"
+  when:
+    - update_routes|default('false')|bool
+    - not (ansible_default_ipv4.network == ( item.cidr | ipaddr('network')))
+
+#Set gateway for each undercloud_osd_cidr subnet. gateway is a member of undercloud_osd_cidr list for every subnet.
+#Each host can belong to only one undercloud_osd_cidr, so there will be only one nw_gw set per host.
+- name: set gateway address for multi subnet
+  set_fact: nw_gw={{ item.gateway }}
+  with_items: "{{ undercloud_osd_cidr }}"
+  when:
+    - update_routes|default('false')|bool
+    - ansible_default_ipv4.network == (item.cidr | ipaddr('network'))
+    - undercloud_osd_cidr is defined
+    - "'ceph_osds' in group_names"
+
+#Add or replace a static route to all undercloud_osd_cidr subnets via the nw_gw of the current subnet.
+- name: add static route for multi subnet
+  command: "/sbin/ip route replace {{ item.cidr }} via {{ nw_gw }}"
+  with_items: "{{ undercloud_osd_cidr }}"
+  when:
+    - update_routes|default('false')|bool
+    - not (ansible_default_ipv4.network == (item.cidr | ipaddr('network')))
+    - undercloud_osd_cidr is defined
```

Comment:
So undercloud_osd_cidr is the flag to decide network separation.
Seems we only consider two cases: undercloud_osd_cidr is defined , undercloud_osd_cidr is undefined.
It's better if we can handle `undercloud_osd_cidr: ~` and `undercloud_osd_cidr: []`
`- undercloud_osd_cidr|default([])` is able to handle the case

If we decide to handle these two cases, related changes are needed in ceph.conf template.